pkg/client/client.go

```diff
@@ -0,0 +1,410 @@
+						// Elastic Agent is requesting us to stop.
+						c.expected = expected.State
+						c.impl.OnStop()
+						continue
+					}
+					if expected.ConfigStateIdx != c.cfgIdx {
+						// Elastic Agent is requesting us to update config.
+						c.cfgLock.Lock()
+						c.cfgIdx = expected.ConfigStateIdx
+						c.cfg = expected.Config
+						c.cfgLock.Unlock()
+						c.impl.OnConfig(expected.Config)
+						continue
+					}
+				}
+			}()
+
+			// observed state check-ins
+			go func() {
+				defer wg.Done()
+
+				var lastSent time.Time
+				var lastSentCfgIdx uint64
+				var lastSentStatus proto.StateObserved_Status
+				var lastSentMessage string
+				for {
+					select {
+					case <-done:
+						return
+					case <-time.After(500 * time.Millisecond):
+					}
+
+					c.cfgLock.RLock()
+					cfgIdx := c.cfgIdx
+					c.cfgLock.RUnlock()
+					c.obsLock.RLock()
+					observed := c.observed
+					observedMsg := c.observedMessage
+					c.obsLock.RUnlock()
+
+					sendMessage := func() {
+						err := checkinClient.Send(&proto.StateObserved{
+							Token:          c.token,
+							ConfigStateIdx: cfgIdx,
+							Status:         observed,
+							Message:        observedMsg,
+						})
+						if err != nil {
+							c.impl.OnError(err)
+							return
+						}
+						lastSent = time.Now()
+						lastSentCfgIdx = cfgIdx
+						lastSentStatus = observed
+						lastSentMessage = observedMsg
+					}
+
+					// On start keep trying to send the initial check-in.
+					if lastSent.IsZero() {
+						sendMessage()
+						continue
+					}
+
+					// Send new status when it has changed.
+					if lastSentCfgIdx != cfgIdx || lastSentStatus != observed || lastSentMessage != observedMsg {
+						sendMessage()
+						continue
+					}
+
+					// Send when more than 30 seconds has passed without any status change.
+					if time.Now().Sub(lastSent) >= c.minCheckTimeout {
+						sendMessage()
+						continue
+					}
+				}
+			}()
+
+			// wait for both send and recv go routines to stop before
+			// starting a new stream.
+			wg.Wait()
+		}
+	}()
+}
+
+// startActions starts the go rountines to send and receive actions
+func (c *Client) startActions() {
+	c.wg.Add(1)
+
+	// results our held outside of the retry loop, because on re-connect
+	// we still want to send the responses that either failed or haven't been
+	// sent back to the agent.
+	actionResults := make(chan *proto.ActionResponse, 100)
+	go func() {
+		defer c.wg.Done()
+		for {
+			select {
+			case <-c.ctx.Done():
+				// stopped
+				return
+			default:
+			}
+
+			actionsClient, err := c.client.Actions(c.ctx)
+			if err != nil {
+				c.impl.OnError(err)
+				continue
+			}
+
+			// initial connection of stream must send the token so
+			// the Elastic Agent knows this clients token.
+			actionResults <- &proto.ActionResponse{
+				Token:  c.token,
+				Id:     "init",
+				Status: proto.ActionResponse_SUCCESS,
+				Result: []byte("{}"),
+			}
+
+			var wg sync.WaitGroup
+			wg.Add(2)
```

Comment:
Reminds me of my comment further above :-)

Reply:
Fixed as well ;-)